src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -624,28 +631,36 @@ void MPS::apply_matrix_internal(const reg_t & qubits, const cmatrix_t &mat,
 void MPS::apply_multi_qubit_gate(const reg_t &qubits,
 				 const cmatrix_t &mat,
 				 bool is_diagonal) {
-  // need to reverse qubits because that is the way they
-  // are defined in the Qiskit interface
-  reg_t reversed_qubits = qubits;
-  std::reverse(reversed_qubits.begin(), reversed_qubits.end()); 
+  // change qubit order in the matrix
+  uint_t nqubits = qubits.size();
+  uint_t sidelen = 1 << nqubits;
+  cmatrix_t new_mat(sidelen, sidelen);
+  for (uint_t i=0; i<sidelen; ++i)
+    for (uint_t j=0; j<sidelen; ++j) {
+      uint_t new_i = reverse_bits(i, nqubits);
+      if (i==j)
+	new_mat(new_i, new_i) = mat(i, i);
+      else {
+	uint_t new_j = reverse_bits(j, nqubits);
+	new_mat(new_i, new_j) = mat(i, j);
+      }
+    }
 
```

Comment:
I'd add braces to the outer loop to avoid possible future errors.

```suggestion
  for (uint_t i=0; i<sidelen; ++i) {
    for (uint_t j=0; j<sidelen; ++j) {
      uint_t new_i = reverse_bits(i, nqubits);
      if (i==j)
        new_mat(new_i, new_i) = mat(i, i);
      else {
        uint_t new_j = reverse_bits(j, nqubits);
        new_mat(new_i, new_j) = mat(i, j);
      }
    }
  }
```